smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/ResolveEndpointMiddlewareGenerator.kt

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package software.amazon.smithy.kotlin.codegen.rendering.endpoints
+
+import software.amazon.smithy.codegen.core.CodegenException
+import software.amazon.smithy.codegen.core.Symbol
+import software.amazon.smithy.kotlin.codegen.KotlinSettings
+import software.amazon.smithy.kotlin.codegen.core.*
+import software.amazon.smithy.kotlin.codegen.model.*
+import software.amazon.smithy.kotlin.codegen.rendering.protocol.ProtocolGenerator
+import software.amazon.smithy.kotlin.codegen.utils.getOrNull
+import software.amazon.smithy.model.Model
+import software.amazon.smithy.model.knowledge.OperationIndex
+import software.amazon.smithy.model.shapes.MemberShape
+import software.amazon.smithy.model.shapes.OperationShape
+import software.amazon.smithy.model.shapes.ServiceShape
+import software.amazon.smithy.rulesengine.language.EndpointRuleSet
+import software.amazon.smithy.rulesengine.language.syntax.parameters.Parameter
+import software.amazon.smithy.rulesengine.language.syntax.parameters.ParameterType
+import software.amazon.smithy.rulesengine.traits.ClientContextParamsTrait
+import software.amazon.smithy.rulesengine.traits.ContextParamTrait
+import software.amazon.smithy.rulesengine.traits.StaticContextParamsTrait
+
+/**
+ * Generates resolver middleware for a specific endpoint provider signature.
+ * The binding of parameter values is generated separately in [EndpointParameterBindingGenerator], and that generated
+ * implementation is plugged in here.
+ */
+class ResolveEndpointMiddlewareGenerator(
```

Comment:
Rather than comment on each individual location I'm going to make one long comment on the resulting generated code.

Taken from S3 it seems we are generating the following middleware and middleware registration:

Current middleware being generated (per service):
```kotlin
public class ResolveEndpointMiddleware<I, O>(
    private val endpointProvider: EndpointProvider,
    private val buildParams: EndpointParameters.Builder.(input: I) -> Unit,
) : InlineMiddleware<I, O> {
    private lateinit var endpoint: Endpoint

    public override fun install(op: SdkHttpOperation<I, O>) {
        op.execution.initialize.intercept { req, next ->
            val params = EndpointParameters { buildParams(req.subject) }
            endpoint = endpointProvider.resolveEndpoint(params)
            next.call(req)
        }

        op.execution.mutate.intercept { req, next ->
            setResolvedEndpoint(req, endpoint)
            next.call(req)
        }
    }
}
```

Example current operation setup/registration:
```kotlin
    override suspend fun abortMultipartUpload(input: AbortMultipartUploadRequest): AbortMultipartUploadResponse {
        ...

        op.install(ResolveEndpointMiddleware<AbortMultipartUploadRequest, AbortMultipartUploadResponse>(config.endpointProvider) { input ->
            bindAwsBuiltins(config)
            bucket = input.bucket
        })
      
        ...
```

-----------------

Couple things stand out to me right away.

The middleware takes a `buildParams` lambda which is passed `req.subject` to bind the input shape members to the endpoint parameters (if there are any). This isn't really necessary though as we have access to the operation input when creating the middleware (and is what you are using already anyway, e.g. `input.bucket` rather than the parameter from the lambda). This relationship is a bit confusing and seems like it's there only because we want to do it in middleware.

I'm almost wondering (and this is just an idea I'm throwing out there I don't even know if I like it yet) if we shouldn't just build the endpoint parameters in the operation setup before middleware registration:

```kotlin
val endpointParameters = EndpointParameters {
    bindBuiltins(config)
    bucket = input.bucket
    ...
}



op.install(ResolveEndpointMiddleware(config.endpointProvider, endpointParameters))

```

`ResolveEndpointMiddleware` could then just be a normal `ModifyRequestMiddleware`

--------

We could update the `EndpointResolver` and `ResolveEndpoint` middleware types to be re-usable such that we don't have to generate a separate resolver type per/client but instead use generics:

something like:
```kotlin
public class ResolveEndpoint<T> {
    private val resolver: EndpointResolver<T>,
    private val parameters: T
): ModiifyRequestMiddleware {
    override suspend fun modifyRequest(req: SdkHttpRequest): SdkHttpRequest {
        val endpoint = resolver.resolve(parameters)
        setResolvedEndpoint(req, endpoint)
        ...
    }   
}
```

Resolver w/generics
```kotlin
public fun EndpointResolver<T> {
    public suspend fun resolve(parameters: T): Endpoint
}
```

The generated client config would then be:

```kotlin
public class Config ... {
     ...
     // specialized to whatever the EndpointParameters type is that IS generated per/service
     public val endpointProvider: EndpointProvider<EndpointParameters> = builder.endpointProvider ?: DefaultEndpointProvider()

}
```

Whether we do this with generics or keep per/service generated resolver type isn't super important it's just one idea to try and re-use as much as possible and not need custom codegen for AWS vs generic smithy models.